Mathlib/Order/Grade.lean

```diff
@@ -0,0 +1,373 @@
+/-
+Copyright (c) 2022 Yaël Dillies, Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Violeta Hernández Palacios, Grayson Burton, Vladimir Ivanov
+
+! This file was ported from Lean 3 source module order.grade
+! leanprover-community/mathlib commit 9003f28797c0664a49e4179487267c494477d853
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Finset.Basic
+import Mathlib.Data.Int.SuccPred
+
+/-!
+# Graded orders
+
+This file defines graded orders, also known as ranked orders.
+
+A `𝕆`-graded order is an order `α` equipped with a distinguished "grade" function `α → 𝕆` which
+should be understood as giving the "height" of the elements. Usual graded orders are `ℕ`-graded,
+cograded orders are `ℕᵒᵈ`-graded, but we can also grade by `ℤ`, and polytopes are naturally
+`Fin n`-graded.
+
+Visually, `Grade ℕ a` is the height of `a` in the Hasse diagram of `α`.
+
+## Main declarations
+
+* `GradeOrder`: Graded order.
+* `GradeMinOrder`: Graded order where minimal elements have minimal grades.
+* `GradeMaxOrder`: Graded order where maximal elements have maximal grades.
+* `GradeBoundedOrder`: Graded order where minimal elements have minimal grades and maximal
+  elements have maximal grades.
+-- PORTING NOTE: not sure whether this should be called `grade` or `Grade`.
+* `Grade`: The grade of an element. Because an order can admit several gradings, the first argument
+  is the order we grade by.
+-- PORTING NOTE: seems to be duplicated.
+* `GradeMaxOrder`: Graded orders with maximal elements. All maximal elements have the same grade.
+-- PORTING NOTE: seems to be duplicated.
+* `MaxGrade`: The maximum grade in a `GradeMaxOrder`.
+-- PORTING NOTE: doesn't seem to exist.
+* `OrderEmbedding.grade`: The grade of an element in a linear order as an order embedding.
+
+## How to grade your order
+
+Here are the translations between common references and our `GradeOrder`:
+* [Stanley][stanley2012] defines a graded order of rank `n` as an order where all maximal chains
+  have "length" `n` (so the number of elements of a chain is `n + 1`). This corresponds to
+  `GradeBoundedOrder (Fin (n + 1)) α`.
+* [Engel][engel1997]'s ranked orders are somewhere between `GradeOrder ℕ α` and
+  `GradeMinOrder ℕ α`, in that he requires `∃ a, IsMin a ∧ Grade ℕ a + 0` rather than
```

Comment:
```suggestion
  `GradeMinOrder ℕ α`, in that he requires `∃ a, IsMin a ∧ Grade ℕ a + 0` rather than
```

this is also some weird docs stuff that I didn't spot, cc @YaelDillies again (sorry, I feel like you're the one most well-versed in this!)

Reply:
Should be
```suggestion
  `GradeMinOrder ℕ α`, in that he requires `∃ a, IsMin a ∧ Grade ℕ a = 0` rather than
```

Reply:
Thanks, do you have any suggestions for the other docs? I guess they should be backported to Lean3, but not sure how high of a priority that is

Reply:
What other docs? The mathlib 3 docs? Feel free to fix it, but it's really not a priority.

Reply:
I left some porting notes in the main declarations that I'm not sure what was meant for and whether the declarations still exist or got moved